config/params/config.go

```diff
@@ -335,22 +347,291 @@ func (b *BeaconChainConfig) ExecutionRequestLimits() enginev1.ExecutionRequestLi
+		MaxBlobsPerBlock: ns.MaxBlobsPerBlock,
+		Epoch:            ns.Epoch,
+		isFork:           ns.isFork,
+	}
 }
 
+type BlobScheduleEntry NetworkScheduleEntry
+
 func (b *BeaconChainConfig) ApplyOptions(opts ...Option) {
 	for _, opt := range opts {
 		opt(b)
 	}
 }
 
+// TODO: this needs to be able to return an error
 // InitializeForkSchedule initializes the schedules forks baked into the config.
 func (b *BeaconChainConfig) InitializeForkSchedule() {
 	// Reset Fork Version Schedule.
 	b.ForkVersionSchedule = configForkSchedule(b)
 	b.ForkVersionNames = configForkNames(b)
+	b.forkSchedule = initForkSchedule(b)
+	b.bpoSchedule = initBPOSchedule(b)
+	combined := b.forkSchedule.merge(b.bpoSchedule)
+	if err := combined.prepare(b); err != nil {
+		log.WithError(err).Error("Failed to prepare network schedule", "error", err)
+	}
+	b.networkSchedule = combined
+}
+
+func LogDigests(b *BeaconChainConfig) {
+	gvr := BeaconConfig().GenesisValidatorsRoot
+	schedule := b.networkSchedule
+	for _, e := range schedule.entries {
+		root, err := computeForkDataRoot(e.ForkVersion, gvr)
+		if err != nil {
+			log.WithField("version", fmt.Sprintf("%#x", e.ForkVersion)).
+				WithField("genesisValidatorsRoot", fmt.Sprintf("%#x", gvr)).
+				WithError(err).Error("Failed to compute fork data root")
+		}
+		fields := logrus.Fields{
+			"forkVersion":      fmt.Sprintf("%#x", e.ForkVersion),
+			"forkDigest":       fmt.Sprintf("%#x", e.ForkDigest),
+			"maxBlobsPerBlock": e.MaxBlobsPerBlock,
+			"epoch":            e.Epoch,
+			"bpoEpoch":         e.BPOEpoch,
+			"isFork":           e.isFork,
+			"forkEnum":         version.String(e.VersionEnum),
+			"sanity":           fmt.Sprintf("%#x", root),
+			"gvr":              fmt.Sprintf("%#x", gvr),
+		}
+		log.WithFields(fields).Debug("Network schedule entry initialized")
+		digests := make([]string, 0, len(schedule.byDigest))
+		for k := range schedule.byDigest {
+			digests = append(digests, fmt.Sprintf("%#x", k))
+		}
+		log.WithField("digest_keys", strings.Join(digests, ", ")).Debug("Digests seen")
+	}
+}
+
+type NetworkSchedule struct {
+	entries   []NetworkScheduleEntry
+	byEpoch   map[primitives.Epoch]*NetworkScheduleEntry
+	byVersion map[[4]byte]*NetworkScheduleEntry
+	byDigest  map[[4]byte]*NetworkScheduleEntry
+}
+
+func newNetworkSchedule(entries []NetworkScheduleEntry) *NetworkSchedule {
+	return &NetworkSchedule{
+		entries:   entries,
+		byEpoch:   make(map[primitives.Epoch]*NetworkScheduleEntry),
+		byVersion: make(map[[4]byte]*NetworkScheduleEntry),
+		byDigest:  make(map[[4]byte]*NetworkScheduleEntry),
+	}
+}
+
+func (ns *NetworkSchedule) epochIdx(epoch primitives.Epoch) int {
+	for i := len(ns.entries) - 1; i >= 0; i-- {
+		if ns.entries[i].Epoch <= epoch {
+			return i
+		}
+	}
+	return -1
+}
+
+func (ns *NetworkSchedule) Next(epoch primitives.Epoch) NetworkScheduleEntry {
+	lastIdx := len(ns.entries) - 1
+	idx := ns.epochIdx(epoch)
+	if idx < 0 {
+		return ns.entries[0]
+	}
+	if idx > lastIdx-1 {
```

Comment:
```suggestion
	if idx == lastIdx {
```
This is equivalent and easier to reason about